src/content_scripts/autocorrect.js

```diff
@@ -0,0 +1,400 @@
+ *
+ * @param {string} str
+ * @returns {number}
+ */
+function countChars(str) {
+    // removing the joiners
+    const split = str.split("\u{200D}");
+    let count = 0;
+
+    for (const s of split) {
+        // removing the variation selectors
+        count += Array.from(s.split(/[\ufe00-\ufe0f]/).join("")).length;
+    }
+
+    return count;
+}
+
+/**
+ * Delete at caret.
+ *
+ * @param {Object} target
+ * @param {string} atext
+ * @returns {void}
+ */
+function deleteCaret(target, atext) {
+    const count = countChars(atext);
+    if (count > 0) {
+        const isSuccess = document.execCommand("delete", false);
+        if (isSuccess) {
+            for (let i = 0; i < count - 1; ++i) {
+                document.execCommand("delete", false);
+            }
+        }
+        // Firefox input and textarea fields: https://bugzilla.mozilla.org/show_bug.cgi?id=1220696
+        else if (typeof target.setRangeText === "function") {
+            const start = target.selectionStart;
+
+            target.selectionStart = start - atext.length;
+            target.selectionEnd = start;
+            target.setRangeText("");
+
+            // Notify any possible listeners of the change
+            const e = document.createEvent("UIEvent");
+            e.initEvent("input", true, false);
+            target.dispatchEvent(e);
+        }
+    }
+}
+
+/**
+ * Convert fractions and constants to Unicode characters.
+ * Adapted from: https://github.com/tdulcet/Tables-and-Graphs/blob/master/graphs.hpp
+ *
+ * @param {number} anumber
+ * @param {number} afraction
+ * @returns {string}
+ */
+function outputLabel(anumber, afraction) {
+    let output = false;
+
+    const number = parseFloat(anumber);
+    let intpart = Math.trunc(number);
+    const fractionpart = afraction ? parseFloat(afraction) : Math.abs(number % 1);
+
+    let strm = "";
+
+    for (const fraction in fractions) {
+        if (Math.abs(fractionpart - fractions[fraction]) < Number.EPSILON) {
+            if (intpart !== 0) {
+                strm += intpart;
+            }
+
+            strm += fraction;
+
+            output = true;
+            break;
+        }
+    }
+
+    if (Math.abs(number) >= Number.EPSILON && !output) {
+        for (const constant in constants) {
+            if (!output && number % constants[constant] === 0) {
+                intpart = number / constants[constant];
+
+                if (intpart === -1) {
+                    strm += "-";
+                } else if (intpart !== 1) {
+                    strm += intpart;
+                }
+
+                strm += constant;
+
+                output = true;
+                break;
+            }
+        }
+    }
+
+    if (!output) {
+        strm += anumber;
+    }
+
+    return strm;
+}
+
+/**
+ * Get first difference index.
+ *
+ * @param {string} a
+ * @param {string} b
+ * @returns {number}
+ */
+function firstDifferenceIndex(a, b) {
```

Comment:
Wait, what exactly is the different to [simple `indexOf`](https://stackoverflow.com/questions/19162845/find-first-index-of-a-string-after-index)?

Reply:
Ah okay, it does sth. else yeah. There is really no built-in JS way for this?

Reply:
Not that I could find when I wrote this.